Preliminary support for Bootstrap 5

taylorthurlow opened the pull request:
With the Bootstrap 5 alpha out today, I thought I'd see how hard it would be to update the gem to support the latest alpha release.

This is a work-in-progress PR, but it does seem to be working, at least with the demo application forms. More work needs to be put into updating parts of the README.

I've also questionably removed some tests that I should probably look into figuring out if they're actually tests that should be removed, or slightly reworked.

One thing that might be worth discussing is Bootstrap 5's removal of the `.form-group` class. The concept is pretty hard-wired into the gem. Even though the class is gone, the "concept" of the form group wrapper is still there, just replaced with the `.mb-3` class to maintain the margin on the div. I don't think it would necessarily be a bad thing to keep using the form group terminology within the code base.

I haven't contributed here before, but any existing maintainers should feel free to push to this branch with changes you feel are necessary. If you'd like to go for a more methodical, less "shotgun search and replace" methodology, I wouldn't be offended if my efforts here were scrapped.

lcreid commented:
I've created a `bootstrap-5` branch, and made a few changes to the `README.md`.

Thanks again for your PR!

taylorthurlow commented:
Switched the base branch. Glad I could be of some help!

taylorthurlow commented:
I noticed that the `bootstrap-5` branch is based on a commit ~8 commits behind the current master - was this intentional? I just rebased my branch onto `bootstrap-5`, it conflicted on a few really trivial things, but I wanted to double check. Cheers!

thimo commented:
Thank you so much for picking up support for Bootstrap 5! I'm already using this in an application I'm working on. 

One thing I noticed is that checkbox labels both get the `form-label`and the `form-check-label` classes. Reading the Bootstrap 5 docs at https://v5.getbootstrap.com/docs/5.0/forms/checks/ it looks like the checkbox labels should not use `form-label`, but only `form-check-label`.

thimo commented:
Another issue: dropdowns (select) should get the class `form-select` instead of `form-control`.

lcreid commented:
Terribly sorry about that. My bad. I have updated the `bootstrap-5` branch. I hope I didn't cause you too much extra work, @taylorthurlow .

@thimo Thanks for the feedback. We're happy to consider pull requests on the `bootstrap-5` branch if you feel up to it. If it's not too much trouble, could you please create two new issues for your two comments, and reference issue #553 in them so we keep track. 

The reason for separate issues is the following: The upgrade from Bootstrap 3 to 4 was a long process, and some of the individual changes turned out to require significant work. It's easier to track the work for each issue separately, rather than try to do one big PR.

Thanks in advance.

taylorthurlow commented:
> Terribly sorry about that. My bad. I have updated the `bootstrap-5` branch. I hope I didn't cause you too much extra work, @taylorthurlow .

No problem at all, it took less than 5 minutes. I kept the old branch alive so I just re-pushed the old branch without the small changes.

I'm not sure what the best approach for getting @thimo's proposed fixes merged here. Would it make sense to merge this PR with its flaws intact, and just continue submitting PRs to the `bootstrap-5` branch? Otherwise I could allow @thimo to push to my branch. Whatever you think the best approach is. I'll mark this PR as ready for review so we can merge it if that's the way to go.

lcreid commented on lib/bootstrap_form/form_group.rb:
```diff
@@ -50,15 +50,23 @@ def form_group_control_class(options)
     end
+
+    def classes_include_gutters?(classes)
+      classes.any? { |c| c =~ /g-\d+/ }
```
Do you need to anchor this regex at the start or end of the string? I'm thinking about a case where the user is using a class name like `setting-5`. (I haven't even looked at the Bootstrap 5 classes yet, so please forgive me if this comment has an obvious answer.)

taylorthurlow replied:
Your concern makes total sense - and you're right, it should be anchored - I'd say both sides.

lcreid commented on lib/bootstrap_form/inputs/check_box.rb:
```diff
@@ -50,40 +50,26 @@ def check_box_value(name, value)
       def check_box_classes(name, options)
-        classes = [options[:class]]
-        classes << (options[:custom] ? "custom-control-input" : "form-check-input")
+        classes = [options[:class], "form-check-input"]
```
Your code appears equivalent to the legacy code with respect to `options[:class]`. However, I'm wondering if a user passed in an array of classes, would the code work? In many places the user can provide an array of classes. (I think it's supported by the underlying Rails helpers.) Maybe we should make this work for the case where someone provides an array of class names in `options[:class]`. What do you think?

taylorthurlow replied:
I just confirmed for myself that the `class` option supports an array of strings. That's cool, didn't know that was supported. Good catch, I'll refactor to account for the option.

taylorthurlow replied:
I'm realizing now that this is why `#flatten` is called in the return value - so this already worked. That said, I'd rather this be clearer so I'll fix it nonetheless.

lcreid commented on lib/bootstrap_form/inputs/radio_button.rb:
```diff
@@ -34,43 +34,25 @@ def radio_button_label(name, value, options)
       def radio_button_classes(name, options)
-        classes = [options[:class]]
-        classes << (options[:custom] ? "custom-control-input" : "form-check-input")
+        classes = [options[:class], "form-check-input"]
```
Same comment as above for check boxes, for `options[:class]` being an array.

lcreid commented on test/bootstrap_checkbox_test.rb:
```diff
@@ -477,63 +477,6 @@ class BootstrapCheckboxTest < ActionView::TestCase
                                                                     :street, checked: collection)
   end
 
```
So the whole "custom" thing is gone? That makes life easier.

taylorthurlow replied:
No kidding, much easier. [Here's the relevant section](https://v5.getbootstrap.com/docs/5.0/migration/#forms) of the Bootstrap 5 migration docs. They're clearly incomplete and missing significant parts, but the linked section says about as much - the custom stuff is gone.

lcreid commented:
@taylorthurlow Apologies for the disorganized nature of my feedback. My paid job doesn't use the gem, so I have to squeeze in time around family and work. Anyway, now that I have my head around @thimo 's comments, I'll leave it up to you if you want to add them to this PR. They seem pretty straightforward, so if you don't mind fixing them in this PR, it would move things along faster.

taylorthurlow commented:
@lcreid Not disorganized at all, thanks for the review. I'll fix the things that @thimo has mentioned tomorrow. 👍

taylorthurlow commented:
I've fixed the issues @thimo brought up. The `form-check-label` issue was a quick fix, but the `form-control` to `form-select` was not.

I feel like the way I fixed it is a little sloppy. The control class is `form-control` by default, and because this was the correct control class for the select box in 4.x, all select fields use the control class definition from `BootstrapForm::FormBuilder#control_class`. In order to override this I copied the `#reverse_merge` technique found here:

https://github.com/taylorthurlow/bootstrap_form/blob/bootstrap5/lib/bootstrap_form/inputs/file_field.rb#L11

And I applied that to all necessary select fields, but for `html_options` instead.

Let me know if you have any suggestions on how to make this a little better.

lcreid commented on lib/bootstrap_form/inputs/base.rb:
```diff
@@ -22,6 +22,7 @@ def bootstrap_select_group(field_name)
           with_field_name = "#{field_name}_with_bootstrap"
           without_field_name = "#{field_name}_without_bootstrap"
           define_method(with_field_name) do |name, options={}, html_options={}|
+            html_options = html_options.reverse_merge(control_class: "form-select")
```
I spent some time looking at this, refreshing my memory about how the code works. The `form_group_builder` method made more sense in earlier versions of Bootstrap, I think. I think your solution is good. I've created an issue #568 to look at whether `form_group_builder` is really serving a purpose now. But your PR is good.

taylorthurlow replied:
Good to hear. The metaprogramming involved made it a little harder to understand what was happening. 👍

thimo commented:
Awesome work, thanks a lot! I've switched to using the new `bootstrap-5` branch and will report back if I run into anything.
